From e943b64e140fcf79e105273c39e1c86e6077f8a1 Mon Sep 17 00:00:00 2001
From: Jan Stancek <jstancek@redhat.com>
Date: Mon, 9 Jun 2014 10:06:12 -0400
Subject: [kernel] pid_namespace: pidns_get() should check task_active_pid_ns() != NULL

Message-id: <41e2a8e73f43e916f3fbdd9f36f304933662109d.1402301509.git.jstancek@redhat.com>
Patchwork-id: 83090
O-Subject: [RHEL7.1 PATCH] pid_namespace: pidns_get() should check task_active_pid_ns() != NULL
Bugzilla: 1106368
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Frantisek Hrbata <fhrbata@redhat.com>
RH-Acked-by: Aristeu S. Rozanski F. <arozansk@redhat.com>
RH-Acked-by: Oleg Nesterov <oleg@redhat.com>

Bugzilla: 1106368
Upstream Status: linux.git commit d23082257d83e4bc89727d5aedee197e907999d2
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7554036
Tested: with usex-1.9-29

commit d23082257d83e4bc89727d5aedee197e907999d2
Author: Oleg Nesterov <oleg@redhat.com>
Date:   Wed Apr 2 17:45:05 2014 +0200

    pidns_get()->get_pid_ns() can hit ns == NULL. This task_struct can't
    go away, but task_active_pid_ns(task) is NULL if release_task(task)
    was already called. Alternatively we could change get_pid_ns(ns) to
    check ns != NULL, but it seems that other callers are fine.

    Signed-off-by: Oleg Nesterov <oleg@redhat.com>
    Cc: Eric W. Biederman ebiederm@xmission.com>
    Cc: stable@kernel.org
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Signed-off-by: Jan Stancek <jstancek@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/pid_namespace.c b/kernel/pid_namespace.c
index 6917e8e..e32703d 100644
--- a/kernel/pid_namespace.c
+++ b/kernel/pid_namespace.c
@@ -312,7 +312,9 @@ static void *pidns_get(struct task_struct *task)
  struct pid_namespace *ns;
 
  rcu_read_lock();
- ns = get_pid_ns(task_active_pid_ns(task));
+ ns = task_active_pid_ns(task);
+ if (ns)
+  get_pid_ns(ns);
  rcu_read_unlock();
 
  return ns;
-- 
1.7.1